File: skellig-test-step-processing/src/main/java/org/skellig/teststep/processing/processor/DefaultTestStepProcessor.java
package org.skellig.teststep.processing.processor;

import org.skellig.teststep.processing.model.TestStep;
import org.skellig.teststep.processing.state.TestScenarioState;
import org.skellig.teststep.processing.validation.TestStepResultValidator;
import org.skellig.teststep.reader.exception.ValidationException;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.Optional;

public class DefaultTestStepProcessor implements TestStepProcessor<TestStep> {

    private List<TestStepProcessor> testStepProcessors;
    private TestScenarioState testScenarioState;
    private TestStepResultValidator validator;

    protected DefaultTestStepProcessor(List<TestStepProcessor> testStepProcessors,
                                       TestScenarioState testScenarioState,
                                       TestStepResultValidator validator) {
        this.testStepProcessors = testStepProcessors;
        this.testScenarioState = testScenarioState;
        this.validator = validator;
    }

    @Override
    public void process(TestStep testStep) {
        Optional<TestStepProcessor> testStepProcessor = testStepProcessors.stream()
                .filter(processor -> testStep.getClass().equals(processor.getTestStepClass()))
                .findFirst();

        testStepProcessor.ifPresent(stepProcessor -> stepProcessor.process(testStep));

        validate(testStep);
    }

    private void validate(TestStep testStep) {
        testStep.getValidationDetails()
                .ifPresent(validationDetails -> {
                    Optional<String> testStepId = validationDetails.getTestStepId();
                    if (testStepId.isPresent()) {
                        Object resultFromOtherTestStep = getResultByTestStepId(testStepId.get());
                        validator.validate(validationDetails.getExpectedResult(), resultFromOtherTestStep);
                    } else {
                        validator.validate(validationDetails.getExpectedResult(), getResultByTestStepId(testStep.getId()));
                    }
                });
    }

    private Object getResultByTestStepId(String testStepId) {
        return testScenarioState.get(testStepId + ".result")
                .orElseThrow(() -> new ValidationException(String.format("Result of the test step '%s' not found", testStepId)));
    }

    @Override
    public Class<TestStep> getTestStepClass() {
        return TestStep.class;
    }

    public static class Builder {

        private List<TestStepProcessor> testStepProcessors;
        private TestScenarioState testScenarioState;
        private TestStepResultValidator validator;

        public Builder() {
            testStepProcessors = new ArrayList<>();
        }

        public Builder withTestStepProcessor(TestStepProcessor<? extends TestStep> testStepProcessor) {
            this.testStepProcessors.add(testStepProcessor);
            return this;
        }

        public Builder withTestScenarioState(TestScenarioState testScenarioState) {
            this.testScenarioState = testScenarioState;
            return this;
        }

        public Builder withValidator(TestStepResultValidator validator) {
            this.validator = validator;
            return this;
        }

        public TestStepProcessor<TestStep> build() {
            Objects.requireNonNull(testScenarioState, "TestScenarioState must be provided");

            return new DefaultTestStepProcessor(testStepProcessors, testScenarioState, validator);
        }
    }
}

File: skellig-test-step-processing/src/main/kotlin/org/skellig/teststep/processing/processor/DefaultTestStepProcessor.kt
package org.skellig.teststep.processing.processor

import org.skellig.teststep.processing.exception.ValidationException
import org.skellig.teststep.processing.model.DefaultTestStep
import org.skellig.teststep.processing.processor.TestStepProcessor.TestStepRunResult
import org.skellig.teststep.processing.state.TestScenarioState
import org.skellig.teststep.processing.util.info
import org.skellig.teststep.processing.util.logTestStepResult
import org.skellig.teststep.processing.util.logger

/**
 * Processes a default test step by running validation of a result from another test step.
 */
internal class DefaultTestStepProcessor private constructor(testScenarioState: TestScenarioState) : ValidatableTestStepProcessor<DefaultTestStep>(testScenarioState) {

    private val log = logger<DefaultTestStepProcessor>()

    override fun process(testStep: DefaultTestStep): TestStepRunResult {
        val testStepRunResult = TestStepRunResult(testStep)
        testScenarioState.set(testStep.getId, testStep)
        log.info(testStep, "Start to process task of test '${testStep.name}' by running validation only")
        validate(testStep, testStepRunResult)

        return testStepRunResult
    }

    private fun validate(testStep: DefaultTestStep, testStepRunResult: TestStepRunResult) {
        var error: RuntimeException? = null
        try {
            validate(testStep)
        } catch (ex: ValidationException) {
            error = ex
        } finally {
            log.logTestStepResult(testStep, "none", error)
            testStepRunResult.notify(null, error)
        }
    }

    /**
     * Validate any result (ex. from other test step).
     * Usually the result from another test step is taken by its `id` property using `get` function,
     * for example:
     *  [get(testA_result).values().size > 0 = true]
     */
    @Throws(ValidationException::class)
    private fun validate(testStep: DefaultTestStep) {
        testStep.validationDetails?.let { validationDetails ->
            validate(testStep, validationDetails, null)
        }
    }

    override fun getTestStepClass(): Class<DefaultTestStep> {
        return DefaultTestStep::class.java
    }

    class Builder : BaseTestStepProcessor.Builder<DefaultTestStep>() {
        override fun build(): TestStepProcessor<DefaultTestStep> {
            return DefaultTestStepProcessor(testScenarioState ?: error("TestScenarioState must be provided"))
        }
    }

}
